File: user-app/lib/models/article.dart
// Article model matching backend ArticleOut schema

class StockImpact {
  final String symbol;
  final String? name;
  final String? sector;
  final String direction; // positive | negative | neutral
  final String effect; // high | medium | low
  final String? reason;

  const StockImpact({
    required this.symbol,
    this.name,
    this.sector,
    required this.direction,
    required this.effect,
    this.reason,
  });

  factory StockImpact.fromJson(Map<String, dynamic> json) {
    final effect = json['effect'] as String? ?? 'low';
    final direction = json['direction'] as String? ?? 'neutral';
    return StockImpact(
      symbol: (json['symbol'] as String? ?? '').toUpperCase().trim(),
      name: json['name'] as String?,
      sector: json['sector'] as String?,
      direction: ['positive', 'negative', 'neutral'].contains(direction)
          ? direction
          : 'neutral',
      effect: ['high', 'medium', 'low'].contains(effect) ? effect : 'low',
      reason: json['reason'] as String?,
    );
  }
}

class Article {
  final String id;
  final String headline;
  final String? content;
  final String? source;
  final String? url;
  final DateTime? publishedAt;
  final DateTime fetchedAt;
  final bool? isFinanciallyRelevant;
  final bool needsReview;
  final String aiStatus;
  final String? credibility;
  final String? geography;
  final double? classificationConfidence;
  final List<Map<String, dynamic>>? impacts;
  final String? summary;
  final String? context;
  final String? impactExplanation;
  final String? keyTakeaway;
  final String? sentiment;
  final List<String> marketsAffected;
  final String? tradeLogic;

  const Article({
    required this.id,
    required this.headline,
    this.content,
    this.source,
    this.url,
    this.publishedAt,
    required this.fetchedAt,
    this.isFinanciallyRelevant,
    required this.needsReview,
    required this.aiStatus,
    this.credibility,
    this.geography,
    this.classificationConfidence,
    this.impacts,
    this.summary,
    this.context,
    this.impactExplanation,
    this.keyTakeaway,
    this.sentiment,
    this.marketsAffected = const [],
    this.tradeLogic,
  });

  factory Article.fromJson(Map<String, dynamic> json) {
    return Article(
      id: json['id'] as String,
      headline: json['headline'] as String,
      content: json['content'] as String?,
      source: json['source'] as String?,
      url: json['url'] as String?,
      publishedAt: json['published_at'] != null
          ? DateTime.tryParse(json['published_at'] as String)
          : null,
      fetchedAt: DateTime.parse(json['fetched_at'] as String),
      isFinanciallyRelevant: json['is_financially_relevant'] as bool?,
      needsReview: json['needs_review'] as bool? ?? false,
      aiStatus: json['ai_status'] as String? ?? 'pending',
      credibility: json['credibility'] as String?,
      geography: json['geography'] as String?,
      classificationConfidence:
          (json['classification_confidence'] as num?)?.toDouble(),
      impacts: (json['impacts'] as List<dynamic>?)
          ?.map((e) => Map<String, dynamic>.from(e as Map))
          .toList(),
      summary: json['summary'] as String?,
      context: json['context'] as String?,
      impactExplanation: json['impact_explanation'] as String?,
      keyTakeaway: json['key_takeaway'] as String?,
      sentiment: json['sentiment'] as String?,
      marketsAffected: (json['markets_affected'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      tradeLogic: json['trade_logic'] as String?,
    );
  }

  /// Returns structured StockImpact objects from the impacts JSON
  List<StockImpact> get stockImpacts {
    if (impacts == null) return [];
    final seen = <String>{};
    final result = <StockImpact>[];
    for (final imp in impacts!) {
      final symbol = (imp['symbol'] as String? ?? '').trim().toUpperCase();
      if (symbol.isNotEmpty && !seen.contains(symbol)) {
        seen.add(symbol);
        result.add(StockImpact.fromJson(imp));
      }
    }
    return result;
  }

  /// Fallback: unique stock symbols from old-style impacts
  List<String> get affectedStocks {
    final si = stockImpacts;
    if (si.isNotEmpty) return si.map((s) => s.symbol).toList();
    if (impacts == null) return [];
    return impacts!
        .expand((i) => (i['stocks'] as List<dynamic>? ?? []).cast<String>())
        .toSet()
        .toList();
  }

  List<String> get sectors {
    if (marketsAffected.isNotEmpty) return marketsAffected;
    if (impacts == null) return [];
    return impacts!
        .map((i) => i['sector'] as String? ?? '')
        .where((s) => s.isNotEmpty)
        .toList();
  }
}

File: backend/app/models/article.py
"""
Article Pydantic + SQLAlchemy models.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel
from sqlalchemy import (
    JSON,
    Boolean,
    Column,
    DateTime,
    Float,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import DeclarativeBase


# ── SQLAlchemy base ───────────────────────────────────────────────────────────


class Base(DeclarativeBase):
    pass


# ── DB Model ──────────────────────────────────────────────────────────────────


class ArticleDB(Base):
    __tablename__ = "articles"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    headline = Column(String(512), nullable=False)
    content = Column(Text, nullable=True)
    source = Column(String(256), nullable=True)
    url = Column(String(2048), nullable=True, unique=True)
    published_at = Column(DateTime(timezone=True), nullable=True)
    fetched_at = Column(DateTime(timezone=True), default=datetime.utcnow)

    # Pipeline flags
    is_financially_relevant = Column(Boolean, nullable=True)
    finance_filter_reason = Column(Text, nullable=True)
    needs_review = Column(Boolean, default=False)
    ai_status = Column(String(32), default="pending")  # pending | done | failed

    # Classification
    credibility = Column(String(32), nullable=True)  # confirmed | opinion
    geography = Column(String(32), nullable=True)  # india | global
    classification_confidence = Column(Float, nullable=True)

    # Impact mapping
    impacts = Column(JSON, nullable=True)  # list of {sector, stocks, direction, effect, symbol}

    # AI enrichment
    summary = Column(Text, nullable=True)
    context = Column(Text, nullable=True)
    impact_explanation = Column(Text, nullable=True)
    key_takeaway = Column(Text, nullable=True)
    sentiment = Column(String(16), nullable=True)  # bullish | bearish | neutral
    markets_affected = Column(JSON, nullable=True)  # list of market/sector strings
    trade_logic = Column(Text, nullable=True)  # Gemini reasoning for the trade

    # Raw
    raw_json = Column(JSON, nullable=True)

    # Vector store sync
    vector_synced = Column(Boolean, default=False)
    vector_synced_at = Column(DateTime(timezone=True), nullable=True)


# ── Pydantic schemas ──────────────────────────────────────────────────────────


class CredibilityEnum(str, Enum):
    confirmed = "confirmed"
    opinion = "opinion"


class GeographyEnum(str, Enum):
    india = "india"
    global_ = "global"


class SentimentEnum(str, Enum):
    bullish = "bullish"
    bearish = "bearish"
    neutral = "neutral"


class ArticleBase(BaseModel):
    headline: str
    content: str | None = None
    source: str | None = None
    url: str | None = None
    published_at: datetime | None = None


class ArticleCreate(ArticleBase):
    raw_json: dict[str, Any] | None = None


class ArticleManualCreate(ArticleBase):
    """For POST /articles/manual"""

    pass


class ArticleOut(ArticleBase):
    id: uuid.UUID
    fetched_at: datetime
    is_financially_relevant: bool | None
    needs_review: bool
    ai_status: str
    credibility: CredibilityEnum | None
    geography: GeographyEnum | None
    classification_confidence: float | None
    impacts: list[dict[str, Any]] | None
    summary: str | None
    context: str | None
    impact_explanation: str | None
    key_takeaway: str | None
    sentiment: SentimentEnum | None
    markets_affected: list[str] | None = None
    trade_logic: str | None = None
    vector_synced: bool = False
    vector_synced_at: datetime | None = None

    model_config = {"from_attributes": True}


class NormalizedArticle(BaseModel):
    """Internal representation passing through the pipeline."""

    headline: str
    content: str | None = None
    source: str | None = None
    url: str | None = None
    published_at: datetime | None = None
    raw_json: dict[str, Any] | None = None
